File: 2021/08/20/main.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        return self.isValidRow(board) and self.isValidCol(board) and self.isValidSquare(board)

    def isValidRow(self, board):
        for row in board:
            if not self.isValid(row): return False
        return True

    def isValidCol(self, board):
        for col in zip(*board):
            if not self.isValid(col): return False
        return True

    def isValidSquare(self, board):
        for i in (0, 3, 6):
            for j in (0, 3, 6):
                s = [board[x][y] for x in range(i,i+3) for y in range(j,j+3)]
                if not self.isValid(s): return False
        return True

    def isValid(self, nums):
        res = [num for num in nums if num != '.']
        return len(res) == len(set(res))

File: 2021/08/20/main.go
package main

func isValidSudoku(board [][]byte) bool {
	for r := 0; r < 9; r++ {
		if !isValidSudokuRow(board, r) {
			return false
		}
	}

	for c := 0; c < 9; c++ {
		if !isValidSudokuCol(board, c) {
			return false
		}
	}

	for i := 0; i < 9; i++ {
		if !isValidSudokuPod(board, i) {
			return false
		}
	}
	return true
}

func isValidSudokuRow(board [][]byte, row int) bool {
	var nums [10]bool
	for col := 0; col < 9; col++ {
		n := convertToNumber(board[row][col])
		if n < 0 {
			continue
		}
		if nums[n] {
			return false
		}
		nums[n] = true
	}
	return true
}

func isValidSudokuCol(board [][]byte, col int) bool {
	var nums [10]bool
	for row := 0; row < 9; row++ {
		n := convertToNumber(board[row][col])
		if n < 0 {
			continue
		}
		if nums[n] {
			return false
		}
		nums[n] = true
	}
	return true
}

func isValidSudokuPod(board [][]byte, pod int) bool {
	var nums [10]bool

	row := (pod / 3) * 3
	col := (pod % 3) * 3

	for drow := 0; drow < 3; drow++ {
		for dcol := 0; dcol < 3; dcol++ {
			n := convertToNumber(board[row+drow][col+dcol])
			if n < 0 {
				continue
			}
			if nums[n] {
				return false
			}
			nums[n] = true
		}
	}
	return true
}

func convertToNumber(b byte) int {
	if b == '.' {
		return -1
	}
	return int(b - '0')
}
